File: backend/src/main/java/com/dreampath/service/chatbot/EmbeddingService.java
package com.dreampath.service.chatbot;

import com.fasterxml.jackson.databind.JsonNode;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

@Service
public class EmbeddingService {

    @Value("${openai.api.key}")
    private String apiKey;

    private final WebClient webClient = WebClient.create();

    public float[] createEmbedding(String text) {

        JsonNode response = webClient.post()
                .uri("https://api.openai.com/v1/embeddings")
                .header("Authorization", "Bearer " + apiKey)
                .header("Content-Type", "application/json")
                .bodyValue("""
                        {
                          "model": "text-embedding-3-small",
                          "input": "%s"
                        }
                        """.formatted(text))
                .retrieve()
                .bodyToMono(JsonNode.class)
                .block();

        JsonNode embedding = response.get("data").get(0).get("embedding");

        float[] vector = new float[embedding.size()];
        for (int i = 0; i < embedding.size(); i++) {
            vector[i] = embedding.get(i).floatValue();
        }

        return vector;
    }
}

File: ai-service/services/vector/embedding_service.py
from typing import List
import os
from openai import OpenAI

from services.rag.pinecone_vector_service import PineconeVectorService

_client = OpenAI(api_key=os.getenv("OPENAI_API_KEY")) if os.getenv("OPENAI_API_KEY") else None


def create_embedding(document: str) -> List[float]:
    if _client is None:
        raise RuntimeError("OPENAI_API_KEY 환경 변수가 설정되어야 합니다.")
    response = _client.embeddings.create(
        model="text-embedding-3-large",
        input=document
    )
    return response.data[0].embedding


async def upload_vector_to_pinecone(vector_id: str, embedding: List[float], metadata: dict):
    service = PineconeVectorService()
    if service.index is None:
        raise RuntimeError("Pinecone index가 초기화되지 않았습니다.")

    service.upsert_vector(vector_id, embedding, metadata)
    return True
